### ISSUE TITLE
Count_occurrence does not work with tokenizer?

### ISSUE THREAD

**Issue by WWWonderer** [google-research/deduplicate-text-datasets#43]
Hello!

Sorry if this is something simple that I missed, but I tried to count occurrences of words in 'rajpurkar/squad' dataset and after downloading data, tokenization and suffix array creation, the binary search from count_occurrence script does not return the correct number of word occurrences (it returns 0 most of the time...). I then digged deeper and found that the problem might be related to tokenization, where the same words can be tokenized differently based on context by a BPE tokenizer. This happens with `GPT2Tokenizer` from huggingface and with `tiktoken` tokenizer also... Below is a script to showcase this difference:

```
from transformers import GPT2Tokenizer
# Initialize the GPT-2 tokenizer
tokenizer = GPT2Tokenizer.from_pretrained("gpt2")

a = "blah blah blah... Architecturally... blah blah"
b = "Architecturally"

# Encode both strings
encoded_a = tokenizer.encode(a)
encoded_b = tokenizer.encode(b)

print("encoded_a: ", encoded_a)
print("encoded_b: ", encoded_b)

# Check if encoded_b is a subsequence of encoded_a
def is_subsequence(sub, main):
    sub_len = len(sub)
    for i in range(len(main) - sub_len + 1):
        if main[i:i + sub_len] == sub:
            return True
    return False

print("Is encoded_b a subsequence of encoded_a?", is_subsequence(encoded_b, encoded_a))

```

The output on my machine is: 
```
encoded_a:  [2436, 993, 33367, 33367, 986, 17340, 20221, 986, 33367, 33367]
encoded_b:  [19895, 5712, 20221]
Is encoded_b a subsequence of encoded_a? False
``` 

This leads to the binary search not finding the correct number of occurrences. Am I missing something? Is there a setting where I can set a BPE tokenizer to always tokenize the same text to the same integers independent  of context?

Thank you!

**Comment by carlini** [comment]
Yeah. This project works in byte-space, not "word"-space.  If you tokenize your dataset, then you'll have to handle things yourself if you want different tokens. (If your dataset isn't that big, you can also just work in byte-space and everything will be fine and easy.)

For your specific example, you'll want to make sure to add an initial space `b = " Architecturally"` and everything should match what you expect. This is because most tokenizers add the space to the following word. So it's not "that" context-dependent, just what the one prior byte was.

If you want to do this properly to do word searching, you'll want to find all first tokens that contain your word and start from there. I don't know exactly how "Architecturally" is tokenized but let's pretend it's "Arch" "itect" "urally". This means you'll want to find all tokens that contain "Arch" which might be "\nArch", " Arch", "Arch", "\tArch". Then run your search for each of those followed by the remaining 2 tokens.

**Comment by WWWonderer** [comment]
Hi Nicholas,

Okay thank you for prompt reply! This helps a lot and indeed by adding a space at the start I was able to find most of the sequences in SQUAD. 

Have a good day!
